substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/graal/snippets/NonSnippetLowerings.java

```diff
@@ -265,6 +265,13 @@ public void lower(FixedNode node, LoweringTool tool) {
                     if (!SubstrateBackend.shouldEmitOnlyIndirectCalls()) {
                         loweredCallTarget = graph.add(new DirectCallTargetNode(parameters.toArray(new ValueNode[parameters.size()]),
                                         callTarget.returnStamp(), signature, targetMethod, callType, invokeKind));
+                    } else if (!targetMethod.hasCodeOffsetInImage()) {
+                        /*
+                         * The target method is not included in the image. This means that it was
+                         * also not needed for the deoptimization entry point. Thus, we are certain
+                         * that this branch will fold away. If not, we will fail later on.
+                         */
+                        return;
```

Comment:
If this assumption fails, this leads to the rather unintuitive error:
```
java.lang.ClassCastException: com.oracle.svm.core.nodes.SubstrateMethodCallTargetNode cannot be cast to org.graalvm.compiler.nodes.LoweredCallTargetNode
```
We could replace the `CallTargetNode` with something like `UnreachableCallTargetNode` to get a bit of a glue why this was not lowered. On the other hand, we should really not run into this error case, so not sure it is worth it.

Reply:
You could do the same trick as in the `implementations.length == 0` case: effectively replace the invocation with a `UnreachedCode` deoptimization.